### PR TITLE
Twemoji integration

### LINKED ISSUE #17063: Research on integrating Twemoji
# Feature Issue

Research on Twemoji Integration in mobile apps.

https://twemoji.twitter.com/

### Acceptance Criteria

- Add a preview in the Quo2 preview

### Notes

The official library doesn't provide support for mobile.

### PR DESCRIPTION
**Pull request by smohamedjavid**
fixes #17063

### Summary

This PR (tries to) integrate Twemoji 🎉 

#### What is Twemoji? (or 𝕏emoji)

Twitter Emoji (Twemoji) is an open-source emoji library, created and owned by 𝕏 (formerly Twitter).
[Official website](https://twemoji.twitter.com) | [GitHub](https://github.com/twitter/twemoji)

From the Twemoji (web/js) script found in their repo, it is evident that the text and HTML nodes are parsed, and if any emoji is found, it is replaced with an image node with a link to the Twemoji equivalent. 

The Twemoji doesn't provide mobile support out of the box. We can not use the web script on mobile. 

#### So, how can we integrate in mobile? 🤔 

React Native support non-text components such (as `View` and `Image`) as children of the `Text` component. So we need to build a parser for mobile to replace the emoji with an Image (used `react-native-fast-image` for caching) that renders the Twemoji equivalent. 

#### Starting with components :iphone:

<table>
<tr><th> Account Avatar </th><th> Channel Avatar</th></tr>
<tr><td></td><td> </td></tr>
</table>

#### Using on messages :speech_balloon:

<table>
<tr><th> Composer </th><th>Message list</th></tr>
<tr><td></td><td> </td></tr>
</table>

<br />

### ⚠️ Pitfalls

- Twemoji supports emojis until Unicode **v14**. At the moment, we have Unicode **v15** which added 31 new emojis such as 🫷🫸🫨. If the user enters these emojis, the app will render them as it is. Twemoji internally (inside 𝕏) added support for Unicode v15, and it's expected to be released 🤞 . 
- As you can infer above screenshot, the composer (`TextInput`) cannot be parsed to render Twemoji. It will render the native (Android/iOS) emoji. 

P.S. Apps that use Twemoji (such as Discord) have rolled out to the public with the above pitfalls

### Review notes

Any suggestions or improvements are appreciated!

### Testing notes

The whole application needs to be tested (we have updated the Text component) to check it works as expected. 

### Platforms

- Android
- iOS

### Steps to test (or play around)

- Open Status
- Navigate to `Quo2 Preview > foundations > typography` (voilá :tada:)

status: ready

### CODE REVIEW

File: src/js/components/twemoji.js [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+        accessible={false}
+        selectable={false}
+        style={{ color: 'transparent', fontSize: 0.1 }}
+      >
+        {' '}
+      </Text>
+      {parsedChildren}
+    </Text>
+  );
+};
```
Comment (smohamedjavid) [inline src/js/components/twemoji.js:246]: #### Reason to add hack/workaround:
There should be a string at the beginning to make sure the image (emoji) is aligned with the baseline of the text.

File: src/quo2/components/common/emoji_hash/view.cljs
Comment (smohamedjavid) [inline src/quo2/components/common/emoji_hash/view.cljs:1]: A helper component to render emoji hash using Twemoji.

I have placed it under `quo2.components.common` as we have a design system component `emoji-hash`, which needs to be implemented.

#### Why not use TwemojiText?

We can not use text (TwemojiText) to parse/display as the text will have a line height greater than the emoji.
From the Figma designs, we can infer that these are images aligned horizontally.

File: src/quo2/components/markdown/text.cljs [hunk 2]
```diff
@@ -31,11 +31,10 @@
 
 (defn- text-view-internal
   [props & children]
-  (let [style (text-style props)]
-    (into [rn/text
-           (merge {:style style}
-                  (dissoc props :style :size :align :weight :color :theme))]
-          children)))
+  (let [style      (text-style props)
+        text-props (merge {:style style}
```
Comment (cammellos) [inline src/quo2/components/markdown/text.cljs:35]:
```
(-> props
     (dissoc :style....)
     (assoc :style style))
```
or any other, but no need to use `merge` here

**Review by cammellos — CHANGES_REQUESTED** [review]
I believe we discussed this before, apologies if you weren't involved in the conversation, and we agreed with the design team product not to use twitter emoji for mvp. 

With regards of this PR, I don't think we should be using this method for emoji, two reasons:

1) It parses the message using regex on clojure/js thread, that's something that it's exploitable in a dos attack and likely will decrease performance in a critical area
2) It remotely fetches emojis, which is something we try not to do as it leaks the user ip address

I personally don't think this should go through, sorry, we can of course have a discussion and happy to hear both sides, though this is definitely something that requires a long discussion :)

**Comment by smohamedjavid** [comment]
> I believe we discussed this before, apologies if you weren't involved in the conversation, and we agreed with the design team product not to use twitter emoji for mvp.

Apologies, I'm not aware of this. If we planned and agreed to not have this for MVP, we can park this PR aside. Revisit it later 👍 .

> With regards of this PR, I don't think we should be using this method for emoji, two reasons:
> 
> 1. It parses the message using regex on clojure/js thread, that's something that it's exploitable in a dos attack and likely will decrease performance in a critical area

Regex is the best option to parse a string at the moment (happy to learn if otherwise). React Native uses the js thread for the (whole) render phase, and UI (native) thread for host views. Maybe we can move these parsing to `status-go` to prevent attacks? On the other hand, emojis are not sensitive data. Sorry, I haven't thought about the attacks in the first place. 

Regards to performance, we can assess the areas which can cause issues.

> 2. It remotely fetches emojis, which is something we try not to do as it leaks the user ip address

Agreed 👍. We can have the emoji image set in our codebase. That would increase the app bundle size by **~3.5 MB**. Initially, planned to add the emoji set resources into the app, but skipped it due to the above reason,

> I personally don't think this should go through, sorry, we can of course have a discussion and happy to hear both sides, though this is definitely something that requires a long discussion :)

Agreed and no worries, Andrea. 😄

**Comment by cammellos** [comment]
> Regex is the best option to parse a string at the moment (happy to learn if otherwise). React Native uses the js thread for the (whole) render phase, and UI (native) thread for host views. Maybe we can move these parsing to status-go to prevent attacks? On the other hand, emojis are not sensitive data. Sorry, I haven't thought about the attacks in the first place.

We parse markdown in status-go, and we could also do emojii in theory (we'd have to check the complexity of parsing emojis that way), though we don't use regex for it, rather we iterate through the string and break at "decisions point" (i.e if you see a `@`, then you check for a mention for the next `n` characters). 
It's a stripped down version of an open source library that we modified for our needs https://github.com/status-im/markdown/tree/master, as we used regexp before and caused dos attacks. 

3.5MB is not a lot, that's definitely worth considering.

Again apologies if this came too late into the dev process, I have should have caught it earlier.

**Comment by smohamedjavid** [comment]
> We parse markdown in status-go, and we could also do emojii in theory (we'd have to check the complexity of parsing emojis that way), though we don't use regex for it, rather we iterate through the string and break at "decisions point" (i.e if you see a `@`, then you check for a mention for the next `n` characters). It's a stripped down version of an open source library that we modified for our needs https://github.com/status-im/markdown/tree/master, as we used regexp before and caused dos attacks.

This seems like a viable solution 💯.  We can explore it further post MVP.

Thanks for the explanation!

> 
> 3.5MB is not a lot, that's definitely worth considering.

Sure, we will add the resources to the app when we revisit it 👍 


> Again apologies if this came too late into the dev process, I have should have caught it earlier.

No worries, Andrea. I will close this PR, and we will revisit it later.

**Pull request closed without merging**